src/engine/x86-64/X86_64Interpreter.v3

```diff
@@ -2440,54 +2440,76 @@ class X86_64InterpreterGen(ic: X86_64InterpreterCode, w: DataWriter) {
+			genReadUleb32(G(cont_id)); // read signature index
+
+			var cont = xenv.tmp1;
+			genPopInto(G(cont));
+			masm.emit_validate_and_consume_cont(xenv.tmp1);
+
 			// move parameter values from %curStack to %cont.top
-			var r_nvals = cont_id;
-			masm.emit_cont_mv(r_vsp, r_cont, r_nvals, r_tmp3, r_tmp4, r_xmm0);
-			asm.call_rel_far(resumeStub);
+			var r_nvals = cont_id; // reuse due to non-overlapping live ranges
+			masm.emit_v3_Instance_heaptypes_r_r(xenv.scratch, xenv.instance);
+			masm.emit_v3_Array_elem_r_rr(ValueKind.REF, xenv.scratch, xenv.scratch, cont_id);
+			masm.emit_get_params_length(r_nvals, xenv.scratch);
+			masm.emit_cont_mv(xenv.vsp, cont, r_nvals, xenv.tmp3, xenv.tmp4, xenv.xmm0);
+
+			// chain {cont to {curStack}, then set {m_curStack} to {cont.top}.
+			var curStack = xenv.tmp4;
+			masm.emit_get_curstack(curStack);
+			masm.emit_v3_set_X86_64Stack_vsp_r_r(curStack, xenv.vsp);
+			masm.emit_v3_set_X86_64Stack_rsp_r_r(curStack, xenv.sp);
+			// reserve space for call addr
+			asm.sub_m_i(G(curStack).plus(offsets.X86_64Stack_rsp), Pointer.SIZE);
+			masm.emit_chain_cont_to_parent(curStack, cont);
+			masm.emit_switch_curStack_to_cont(cont);
+
+			asm.call_rel_far(stub_resume);
 			/* === CHILD STACK RUNNING === */
-			genOnResumeFinish(false);
+			genOnResumeFinish();
 			endHandler();
-			asm.bind(resumeStub); { genResumeStub(r_cont, r_tmp0, r_tmp3, false); }
-		}
-		bindHandler(Opcode.RESUME_THROW); {
-			var resumeStub = X86_64Label.new();
-			var cont_id = r_tmp0, ip_store = r_tmp1, tag_index = r_tmp4;
-			computeCurIpForTrap(-1);
-			computePcFromCurIp();
-			// read info and rewind %ip (need %ip at start of opcode during call)
-			asm.movq_r_r(ip_store, r_ip);
-			genReadUleb32(cont_id); // read signature index
-			genReadUleb32(tag_index); // read tag index
-			asm.movq_r_r(r_ip, ip_store);
-			saveCallerIVars();
-			var r_cont = r_tmp1;
-			genResume(cont_id, r_cont, r_tmp3);
-			var r_tag = r_tmp3;
-			asm.movq_r_m(r_scratch, r_instance.plus(offsets.Instance_tags));
-			asm.movq_r_m(r_tag, r_scratch.plusR(tag_index, offsets.REF_SIZE, offsets.Array_contents));
-			masm.emit_get_curstack(xenv.tmp4);
-			saveCallerIVars();
-			callRuntime(refRuntimeCall(X86_64RT.runtime_process_resume_throw), [r_tmp4, r_instance, r_cont, r_tag], false);
-			asm.movq_r_r(r_cont, r_runtime_ret0);
-			restoreCallerIVars();
-			asm.call_rel_far(resumeStub);
-			/* === CHILD STACK RUNNING === */
-			genOnResumeFinish(true);
-			endHandler();
-
-			asm.bind(resumeStub); { genResumeStub(r_cont, r_tmp0, r_tmp3, true); }
+			asm.bind(stub_resume); {
+				// load new stack pointers
+				masm.emit_get_curstack(curStack);
+				masm.emit_v3_X86_64Stack_vsp_r_r(xenv.vsp, curStack);
+				masm.emit_v3_X86_64Stack_rsp_r_r(xenv.sp, curStack);
+				// pop address from new stack and go to it
+				masm.emit_pop_r(ValueKind.REF, xenv.scratch);
+				masm.emit_jump_r(xenv.scratch);
+			}
 		}
+		// bindHandler(Opcode.RESUME_THROW); {
```

Comment:
Is this code dead now?

Reply:
There was a bug in the old implementation that causes new tests to fail. I plan to implement the new stack walk first and base this handler on top of that in a separate PR (since the handler would change a lot after the rewrite) instead of fixing the current implementation. What do you think?

Reply:
Sounds ok.